src/gather_classes.rs

```diff
@@ -257,80 +259,148 @@ fn update_report(
             debug!("skipping transaction: `{transaction}`");
             continue;
         }
-        let calls = get_calls(root_call).context("Failed to get calls from root_call")?;
-        for call in calls {
-            let maybe_kvp = get_tuple_from_call(call);
-            match maybe_kvp {
-                Ok(kvp) => {
-                    report.add_fields(kvp.1, kvp.0);
-                    updates += 1;
-                }
-                Err(err) => {
-                    debug!("Failed to extract tuple with error: {err}");
-                }
-            }
+        let calls =
+            get_calls_with_count(root_call).context("Failed to get calls from root_call")?;
+        for ((entry_point, class_hash), count) in calls {
+            report.update_count(class_hash, entry_point, count);
+            updates += count;
         }
     }
     assert!(report.blocks.insert(block_num));
     Ok(updates)
 }
 
-// Converts recursive calls to value.get("calls") to a boxed iterator.
-fn get_calls<'a>(obj: &'a Value) -> Result<Box<dyn Iterator<Item = &Value> + 'a>, anyhow::Error> {
+/// Merges two lists of calls into a single list of calls, taking the sum of matching keys.
+/// TODO: should this take the max instead?
+fn merge_calls_with_count(
+    base_calls: Vec<CallWithCount>,
+    native_calls: Vec<CallWithCount>,
+) -> Vec<CallWithCount> {
+    let mut merged_map: HashMap<CallKey, usize> = HashMap::new();
+
+    for (key, count) in base_calls.into_iter().chain(native_calls.into_iter()) {
+        merged_map
+            .entry(key)
+            .and_modify(|existing_count| *existing_count += count)
+            .or_insert(count);
+    }
+
+    merged_map.into_iter().collect()
+}
+
+/// Converts recursive calls to value.get("calls") to a boxed iterator.
+fn get_calls_with_count(obj: &Value) -> Result<Vec<CallWithCount>, anyhow::Error> {
```

Comment:
Will change back to Iter once logic is correct

Reply:
You don't have to keep iter. In my opinion is a bit of an unnecessary abstraction, you can return what you think is best.